src/uct/cuda/cuda_ipc/cuda_ipc_md.c

```diff
@@ -339,52 +355,50 @@ uct_cuda_ipc_md_open(uct_component_t *component, const char *md_name,
-
-    com     = ucs_derived_of(md->super.component, uct_cuda_ipc_component_t);
-    com->md = md;
-    *md_p   = &md->super;
+    md->ops       = &md_ops;
+    md->component = &uct_cuda_ipc_component;
+    *md_p         = md;
     return UCS_OK;
 }
 
-uct_cuda_ipc_component_t uct_cuda_ipc_component = {
-    .super = {
-        .query_md_resources = uct_cuda_base_query_md_resources,
-        .md_open            = uct_cuda_ipc_md_open,
-        .cm_open            = ucs_empty_function_return_unsupported,
-        .rkey_unpack        = uct_cuda_ipc_rkey_unpack,
-        .rkey_ptr           = ucs_empty_function_return_unsupported,
-        .rkey_release       = uct_cuda_ipc_rkey_release,
-        .rkey_compare       = uct_base_rkey_compare,
-        .name               = "cuda_ipc",
-        .md_config          = {
-            .name           = "Cuda-IPC memory domain",
-            .prefix         = "CUDA_IPC_",
-            .table          = uct_cuda_ipc_md_config_table,
-            .size           = sizeof(uct_cuda_ipc_md_config_t),
-        },
-        .cm_config          = UCS_CONFIG_EMPTY_GLOBAL_LIST_ENTRY,
-        .tl_list            = UCT_COMPONENT_TL_LIST_INITIALIZER(&uct_cuda_ipc_component.super),
-        .flags              = 0,
-        .md_vfs_init        =
-                (uct_component_md_vfs_init_func_t)ucs_empty_function
+uct_component_t uct_cuda_ipc_component = {
+    .query_md_resources = uct_cuda_base_query_md_resources,
+    .md_open            = uct_cuda_ipc_md_open,
+    .cm_open            = ucs_empty_function_return_unsupported,
+    .rkey_unpack        = uct_cuda_ipc_rkey_unpack,
+    .rkey_ptr           = ucs_empty_function_return_unsupported,
+    .rkey_release       = uct_cuda_ipc_rkey_release,
+    .rkey_compare       = uct_base_rkey_compare,
+    .name               = "cuda_ipc",
+    .md_config          = {
+        .name           = "Cuda-IPC memory domain",
+        .prefix         = "CUDA_IPC_",
+        .table          = uct_cuda_ipc_md_config_table,
+        .size           = sizeof(uct_cuda_ipc_md_config_t),
     },
-    .md                     = NULL,
+    .cm_config          = UCS_CONFIG_EMPTY_GLOBAL_LIST_ENTRY,
+    .tl_list            = UCT_COMPONENT_TL_LIST_INITIALIZER(&uct_cuda_ipc_component),
+    .flags              = 0,
+    .md_vfs_init        = (uct_component_md_vfs_init_func_t)ucs_empty_function
 };
-UCT_COMPONENT_REGISTER(&uct_cuda_ipc_component.super);
+UCT_COMPONENT_REGISTER(&uct_cuda_ipc_component);
 
+UCS_STATIC_INIT {
+    ucs_recursive_spinlock_init(&uct_cuda_ipc_uuid_cache.lock, 0);
+    kh_init_inplace(cuda_ipc_uuid_hash, &uct_cuda_ipc_uuid_cache.hash);
+}
+
+UCS_STATIC_CLEANUP {
+    uct_cuda_ipc_dev_cache_t *cache;
+
+    kh_foreach_value(&uct_cuda_ipc_uuid_cache.hash, cache, {
+        free(cache);
+    })
+    kh_destroy_inplace(cuda_ipc_uuid_hash, &uct_cuda_ipc_uuid_cache.hash);
+    ucs_recursive_spinlock_destroy(&uct_cuda_ipc_uuid_cache.lock);
+}
```

Comment:
Right, I see the potential problem, these calls may be thrown away by a linker. Do we have the proper way to handle it?

I ask because I found this pattern used in cuda_ipc_cache.c, but it's also used in other places (event.c, vfs_obj.c), so my assumption was that it's a valid thing to use. If it's not the case, maybe we should come up with approach that handles all those cases?

Reply:
use KHASH_STATIC_INITIALIZER, make the lock pthread_mutex_t, and use PTHREAD_MUTEX_INITIALIZER

Reply:
ok